File: plt-att-speech-objc/plt-att-speech-objc/ViewController.m
//
//  ViewController.m
//  plt-att-speech-objc
//
//  Created by Morgan Davis on 11/2/15.
//  Copyright © 2015 Plantronics. All rights reserved.
//

#import "ViewController.h"
@import AVFoundation;


#define DOCUMENTS_PATH  [NSHomeDirectory() stringByAppendingPathComponent:@"Documents"]
#define RECORDING_PATH  [DOCUMENTS_PATH stringByAppendingPathComponent:@"recording.caf"]
#define RECORDING_URL   [NSURL fileURLWithPath:RECORDING_PATH]


@interface ViewController () <AVAudioRecorderDelegate, AVAudioPlayerDelegate>

- (IBAction)listenStopButton:(UIButton *)sender;
- (void)setupAudioSession;
- (void)startRecording;
- (void)stopRecording;
- (void)playAudioData:(NSData *)audioData;

@property (nonatomic, retain) AVAudioRecorder       *recorder;
@property (nonatomic, retain) AVAudioPlayer         *player;

@end


@implementation ViewController

#pragma mark - Public

- (IBAction)listenStopButton:(UIButton *)sender
{
    NSLog(@"listenStopButton:");
    
    if (!self.recorder || !self.recorder.isRecording) {
        // start recording
        [self setupAudioSession];
        [self startRecording];
        [sender setTitle:@"Stop Listening" forState:UIControlStateNormal];
    }
    else {
        // stop recording
        [self stopRecording];
        [sender setTitle:@"Start Listening" forState:UIControlStateNormal];
    }
}

#pragma mark - Private

- (void)setupAudioSession
{
    NSLog(@"setupAudioSession");
    
    NSError *err = nil;
    AVAudioSession *audioSession = [AVAudioSession sharedInstance];
    
    // setup audio session
    
    [audioSession setMode:AVAudioSessionModeDefault error:&err];
    if (err) { NSLog(@"Error configuring audio session: %@", audioSession); }
    
    [audioSession setCategory:AVAudioSessionCategoryPlayAndRecord withOptions:AVAudioSessionCategoryOptionAllowBluetooth error:&err];
    if (err) { NSLog(@"Error configuring audio session: %@", audioSession); }
    
    [audioSession setActive:YES error:&err];
    if (err) {  NSLog(@"Error activiting audio session: %@", audioSession); }

    // set BT headset as input (also sets output)
    
    for (AVAudioSessionPortDescription *route in audioSession.availableInputs) {
        if ([route.portType isEqualToString:AVAudioSessionPortBluetoothHFP] ) {
            NSLog(@"Setting input route to: %@", route.portType);
            err = nil;
            [audioSession setPreferredInput:route error:&err];
            if (err) { NSLog(@"Error setting input audio route: %@", err); }
            break;
        }
    }
}

- (void)startRecording
{
    NSLog(@"startRecording");
    
    NSError *err = nil;
    NSFileManager *fm = [NSFileManager defaultManager];
    [fm removeItemAtPath:RECORDING_PATH error:&err];
    if (err) { NSLog(@"Error removing previous recording: %@", err); }

    NSMutableDictionary *settings = [NSMutableDictionary dictionary];
    settings[AVFormatIDKey] = @(kAudioFormatLinearPCM);
    settings[AVSampleRateKey] = @(44100.0);
    settings[AVNumberOfChannelsKey] = @(1);
    settings[AVLinearPCMBitDepthKey] = @(16);
    settings[AVLinearPCMIsBigEndianKey] = @(NO);
    settings[AVLinearPCMIsFloatKey] = @(NO);
    
    NSURL *url = [NSURL fileURLWithPath:RECORDING_PATH];
    err = nil;
    self.recorder = [[AVAudioRecorder alloc] initWithURL:url settings:settings error:&err];
    if (err) {
        NSLog(@"Error initializing audio recorder: %@", err);
    }
    else {
        self.recorder.delegate = self;
        [self.recorder prepareToRecord];
        [self.recorder record];
    }
}

- (void)stopRecording
{
    NSLog(@"stopRecording");
    
    [self.recorder stop];
}

- (void)playAudioData:(NSData *)audioData
{
    NSLog(@"playAudioData: %lu", (unsigned long)audioData.length);
    
    if (self.player) {
        NSLog(@"Killing old player");
        [self.player stop];
        self.player = nil;
    }
    
    NSError *err = nil;
    self.player = [[AVAudioPlayer alloc] initWithData:audioData error:&err];
    if (err) {
        NSLog(@"Error initializing audio player: %@ %ld %@", err.domain, (long)err.code, err.userInfo);
    }
    else {
        NSLog(@"audioData: %ld", audioData.length);
        self.player.delegate = self;
        self.player.volume = 1.0;
        [self.player prepareToPlay];
        [self.player play];
    }
}

#pragma mark - AVAudioRecorderDelegate

- (void)audioRecorderDidFinishRecording:(AVAudioRecorder *)recorder successfully:(BOOL)flag
{
    NSLog (@"audioRecorderDidFinishRecording: %@ successfully: %@", recorder, (flag ? @"YES" : @"NO"));
    
    NSError *err = nil;
    NSData *audioData = [NSData dataWithContentsOfURL:RECORDING_URL options:0 error:&err];
    if (!audioData) {
        NSLog(@"Error getting audio data from file: %@", err);
    }
    else {
        [self playAudioData:audioData];
    }
}

#pragma mark - AVAudioPlayerDelegate

- (void)audioPlayerDidFinishPlaying:(AVAudioPlayer *)player successfully:(BOOL)flag
{
    NSLog(@"audioPlayerDidFinishPlaying: %@ successfully: %@", player, (flag ? @"YES" : @"NO"));
}

- (void)audioPlayerDecodeErrorDidOccur:(AVAudioPlayer *)player error:(NSError * __nullable)error
{
    NSLog(@"audioPlayerDecodeErrorDidOccur: %@ error: %@", player, error);
}

- (void)audioPlayerBeginInterruption:(AVAudioPlayer *)player
{
    NSLog(@"audioPlayerBeginInterruption: %@", player);
}

- (void)audioPlayerEndInterruption:(AVAudioPlayer *)player withOptions:(NSUInteger)flags
{
    NSLog(@"audioPlayerEndInterruption: %@ withOptions: %lu", player, (unsigned long)flags);
}

- (void)audioPlayerEndInterruption:(AVAudioPlayer *)player withFlags:(NSUInteger)flags
{
    NSLog(@"audioPlayerEndInterruption: %@ withFlags: %lu", player, (unsigned long)flags);
}

- (void)audioPlayerEndInterruption:(AVAudioPlayer *)player
{
    NSLog(@"audioPlayerEndInterruption: %@", player);
}

#pragma mark - UIViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

@end

File: plt-att-speech-swift/plt-att-speech-swift/ViewController.swift
//
//  ViewController.swift
//  plt-att-speech-swift
//
//  Created by Morgan Davis on 11/2/15.
//  Copyright © 2015 Plantronics. All rights reserved.
//

import UIKit
import AVFoundation


public class ViewController: UIViewController, AVAudioRecorderDelegate, AVAudioPlayerDelegate {
    
    /******************************************************************************************************
    MARK:   Constants
    ******************************************************************************************************/
    
    private let RECORDING_URL =     NSFileManager.defaultManager()
        .URLsForDirectory(.DocumentDirectory, inDomains: [.UserDomainMask])[0]
        .URLByAppendingPathComponent("recording.caf")
    
    /******************************************************************************************************
    MARK:   Properties
    ******************************************************************************************************/
    
    private var recorder:           AVAudioRecorder?
    private var player:             AVAudioPlayer?
    
    /******************************************************************************************************
    MARK:   Public
    ******************************************************************************************************/
    
    @IBAction func listenStopButton(sender: UIButton) {
        NSLog("listenStopButton(%@)", sender)
        
        if recorder == nil || !recorder!.recording {
            // start recording
            setupAudioSession()
            startRecording()
            sender.setTitle("Stop Listening", forState: .Normal)
        }
        else {
            // stop recording
            stopRecording()
            sender.setTitle("Start Listening", forState: .Normal)
        }
    }
    
    /******************************************************************************************************
    MARK:   Private
    ******************************************************************************************************/
    
    private func setupAudioSession() {
        NSLog("setupAudioSession()")
        
        // setup audio session
        
        let audioSession = AVAudioSession.sharedInstance()
        
        do {
            try audioSession.setMode(AVAudioSessionModeDefault)
            try audioSession.setCategory(AVAudioSessionCategoryPlayAndRecord, withOptions: [.AllowBluetooth])
            try audioSession.setActive(true)
        }
        catch {
            NSLog("Error setting up audio session.")
        }
        
        // set BT headset as input (also sets output)
        
        if let inputRoutes = audioSession.availableInputs {
            for route in inputRoutes {
                if route.portType == AVAudioSessionPortBluetoothHFP {
                    NSLog("Setting input route to: %@", route.portType)
                    do {
                        try audioSession.setPreferredInput(route)
                    }
                    catch {
                        NSLog("Error setting audio input route.")
                    }
                }
            }
        }
        else {
            NSLog("No input audio routes found.")
        }
    }
    
    private func startRecording() {
        NSLog("startRecording()")
        
        let fileManager = NSFileManager.defaultManager()
        try? fileManager.removeItemAtURL(RECORDING_URL)
        
        var settings = [String : AnyObject]()
        settings[AVFormatIDKey] = NSNumber(unsignedInt: kAudioFormatLinearPCM)
        settings[AVSampleRateKey] = 44100.0
        settings[AVNumberOfChannelsKey] = 1
        settings[AVLinearPCMBitDepthKey] = 16
        settings[AVLinearPCMIsBigEndianKey] = false
        settings[AVLinearPCMIsFloatKey] = false
        
        do {
            try recorder = AVAudioRecorder(URL: RECORDING_URL, settings: settings)
            recorder?.delegate = self
            recorder?.prepareToRecord()
            recorder?.record()
        }
        catch {
            NSLog("Error creating audio recorder.")
        }
    }
    
    private func stopRecording() {
        NSLog("stopRecording()")
        
        recorder?.stop()
    }
    
    private func playAudioData(audioData: NSData) {
        NSLog("playAudioData(%d)", audioData.length)
        
        if player != nil {
            NSLog("Killing old player")
            player?.stop()
            player = nil
        }
        
        do {
            try player = AVAudioPlayer(data: audioData)
            player?.delegate = self
            player?.volume = 1.0
            player?.prepareToPlay()
            player?.play()
        }
        catch {
            NSLog("Exceptioon initializing AVAudioPlayer.")
        }
    }
    
    /******************************************************************************************************
    MARK:   AVAudioRecorderDelegate
    ******************************************************************************************************/
    
    public func audioRecorderDidFinishRecording(recorder: AVAudioRecorder, successfully flag: Bool) {
        NSLog("audioRecorderDidFinishRecording(%@, successfully: %@)", recorder, (flag ? "true" : "false"))
        
        if flag {
            if let audioData = NSData(contentsOfURL: RECORDING_URL) {
                playAudioData(audioData)
            }
            else {
                NSLog("Error getting audio data.")
            }
        }
        else {
            NSLog("Error recording audio.")
        }
    }
    
    /******************************************************************************************************
    MARK:   AVAudioPlayerDelegate
    ******************************************************************************************************/
    
    public func audioPlayerDidFinishPlaying(player: AVAudioPlayer, successfully flag: Bool) {
        NSLog("audioPlayerDidFinishPlaying(%@, successfully: %@)", player, (flag ? "true" : "false"))
    }
    
    public func audioPlayerDecodeErrorDidOccur(player: AVAudioPlayer, error: NSError?) {
        NSLog("audioPlayerDecodeErrorDidOccur(%@, error: %@)", player, (error == nil ? "nil" : error!))
    }
    
    public func audioPlayerBeginInterruption(player: AVAudioPlayer) {
        NSLog("audioPlayerBeginInterruption(%@)", player)
    }
    
    public func audioPlayerEndInterruption(player: AVAudioPlayer) {
        NSLog("audioPlayerEndInterruption(%@)", player)
    }
    
    public func audioPlayerEndInterruption(player: AVAudioPlayer, withFlags flags: Int) {
        NSLog("audioPlayerEndInterruption(%@, withFlags: %X)", player, flags)
    }
    
    public func audioPlayerEndInterruption(player: AVAudioPlayer, withOptions flags: Int) {
        NSLog("audioPlayerEndInterruption(%@, withFlags: %X)", player, flags)
    }
    
    /******************************************************************************************************
    MARK:   UIViewController
    ******************************************************************************************************/
    
    override public func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override public func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
}
